### ISSUE TITLE
Separate recorder into calculation and writer parts

### ISSUE THREAD

**Issue by snorfalorpagus** [pywr/pywr#182]
We have classes that record:

- Flow
- Storage
- Level
- Arbitrary `Parameter` values (#181)
- Aggregate statistics (e.g. total deficit)

We have classes that can write:

- Numpy array
- CSV
- HDF5

The current recorder API couples the type of information being recorded and the method for recording it, thus we end up with classes such as `NumpyArrayNodeRecorder` which is different from `NumpyArrayStorageRecorder`. It also means that some combinations have not been implemented.

I propose we modify the API so that the storage method (the "writer") becomes a property of a more general recorder. e.g. a `NodeRecorder` would have a `writer` property/kwarg of type `NumpyArray` (or similar). The main recorder class would be responsible for working out the size of the array required and calculating the value to be stored, while the writer is responsible only for storing the data (on disk / in memory / over a network).

The writer doesn't just write data (perhaps it needs another name?) as it also needs to be able to read data to support the `to_dataframe` method.

**Comment by snorfalorpagus** [comment]
To explain a little further my issue with the current setup...

`NumpyArrayNodeRecorder` inherits from `NodeRecorder`. But the actual code to get the value to recorder isn't in `NodeRecoder` - this means that any other formats need to re-implement something we've already done. It also means that if we add a new format for recording data we need to write a whole new set of classes, rather than just passing the new format as an argument.

I'm thinking of something like this:

```
rec = NodeRecoder(node, format="numpy")
```

Or for aggregate recorders (where you have multiple recorders in the same table, e.g. CSV or HDF5):

```
rec1 = NodeRecorder(node1)
rec2 = NodeRecorder(node2)
rec = CSVRecorder(recorders=[rec1, rec2])
```

Equally it should be valid to specify a CSV to record a single node:

```
rec = NodeRecorder(node1, format="csv")
```

**Comment by snorfalorpagus** [comment]
I was thinking something like this:

https://gist.github.com/snorfalorpagus/ee7541bf150b55db5ea21b12e0b23e88

**Comment by jetuk** [comment]
You would want to be able to manipulate the backend objects in more complex examples though?

**Comment by snorfalorpagus** [comment]
Yes it needs more work. I haven't thought much about scenarios, or data that doesn't have the (timesteps, scenarios) shape.

**Comment by jetuk** [comment]
OK started hacking on this.

How about `EventRecorder`, `TimeseriesRecorder` and `ValueRecorder` each with their own specific subclasses, but how they have have corresponding backends that are called differently.

**Comment by snorfalorpagus** [comment]
So we'd have `FlowRecorder` and `VolumeRecorder` which is are subclasses of `TimeseriesRecorder`, and wrote data to a `TimeseriesRecorderBackend` (e.g. `TimeseriesRecorderNumpyBackend`)? I like the first part of that, but don't especially like the idea of having duplicates of the backends.

**Comment by jetuk** [comment]
Unfortunately  the nature of the data for event and timeseries records are different, so I don't think they can be shared.

**Comment by jetuk** [comment]
Agreed to delay this change for a bit.

**Comment by jetuk** [comment]
We never resolved that the nature of the data is related to the writer backend, hence the separation of concerns is not so easy in practice.

**Comment by jetuk** [comment]
We recently agreed to progress this idea. The PoC here needs a review.

**Comment by jetuk** [comment]
Some initial thoughts from me. A common use case for use for us to record a metric for all timesteps, then want some aggregated metrics of it. We provide the `temporal_agg_func` to the `NumpyArray*Recorder`s to achieve this. They are currently limited by only allowing one aggregation function each. This means if you want some other metrics you must use two recorders and duplicate the memory requirement.

For any new system it would be good to support the following:
1. Define a "metric" to be recorded (e.g. deficit). The metric produces this number from the model every time-step.
2. Record a metric in memory (i.e. a memory backend), and possibly separately to disk (i.e. tables backend).
3. Optionally apply multiple aggregation (and resampling?) functions to the in-memory data (e.g. min, max, percentile), or we have some other aggregated metrics (e.g. total) that can work directly with the metric in (1).

Values from (3) can be used as objectives or constraints in the optimisation wrappers.

**Comment by Batch21** [comment]
So could this new system involve splitting recorders into 3 parts; metrics, aggregations, and writers? I think this makes sense for some cases, for example, the duration recorders are basically `NumpyArray*Recorders` that do more complex aggregations at the end. What about more complex recorders, such as the ones tracking events or using rolling aggregations - would these be metrics that could then be further aggregated? Just trying to get my head around how this system could be applied across all the recorders.
